### DISCUSSION TITLE
How to use Background Tasks with ContextManagers in a Dependency-like Manner?

### DISCUSSION THREAD

**Discussion by thurse93** [fastapi/fastapi#3513] [Questions] [answered]
Hi!

I'm fairly new to FastAPI but I like it alot so far. Im working my way through to the more advanced stuff now; currently I'm trying to send E-mails as a background task. For this I want to use the internal Mail Client from our company, which is a context manager and works something like this:

```python
with MailClient(credentials) as mail_client:
    mail.send(to="alice@example.com", subject="Hello world", message="some message")
```

However im struggling how put together all the information I got from https://fastapi.tiangolo.com/tutorial/background-tasks/ and https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/ for my goal.

What I'd like to achieve is something like that:

```python
@app.get("/send")
async def(mailer: Mailer = Depends()):
    mailer.send(to="alice@example.com", subject="Hello world", message="some message")
```
The Mailers send method should have the same signature as MailClient (or just use kwargs). Internally it should call add_task to invoke the send-method of the MailClient in background. Additionally the Mailer should open and close the connection to the smtp-Server (this is what the MailClient does in its \_\_enter__ and \_\_exit__ methods) when the request is processed.

Is this even possible?

Regards,

**Comment by thurse93** [comment, marked as answer]
I came up with this, but i dont know if it's the correct way of doing it:

```python
from mycompany import MailClient

def _client():
    with MailClient({'username': 'john', 'password': 'secretsauce'}) as dep:
        yield dep


class Mail:
    def __init__(self, bt, client):
        self.background_tasks = bt
        self.client= client

    @staticmethod
    def inject(background_tasks: BackgroundTasks, client=Depends(_client)):
        return Mail(background_tasks, client)

    def send(self, *args, **kwargs):
        self.background_tasks.add_task(self.client.send, *args, **kwargs)
```

Now i can just

```python
from tasks import Mailer

app = FastAPI()

@app.get("/send")
async def send_mail(mail: Mail = Depends(Mail.inject)):
    mail.send(message="...")
```

**Reply by ghandic** [reply]
Using dependency injection like this works well, @lesnik512 's answer is interesting too as it looks to have less coupling to FastAPI which may make it easier to test? 

Which solution did you end up using? Could you mark it as answered to help out others coming for a solution? 😄

**Reply by thurse93** [reply]
We ended up using our approach since it explicitly relies on background tasks and we don't mind the coupling (it's part of a library that requires FastAPI anyway), even though @lesnik512 solution woud be preferrable for final applications.

**Comment by lesnik512** [comment]
I would use something like this:

```python
async def send_mail():
    try:
        yield
    finally:
        # sending mail
        # this part will be executed after the response, so it will be like background task, I think

@app.get("/send", dependencies=[Depends(send_mail)])
async def send_mail():
    pass
```
